pkg/jobWorker/outputReadCloser.go

```diff
@@ -0,0 +1,75 @@
+package jobWorker
+
+import (
+	"errors"
+)
+
+var (
+	ErrOutputMissing = errors.New("OutputReader's CommandOutput is nil")
+	ErrReaderClosed  = errors.New("OutputReader is closed")
+)
+
+// OutputReadCloser implements io.ReadCloser interface to read from the provided CommandOutput
+//
+//	and close output if it is no longer need it .
+type OutputReadCloser struct {
+	output *CommandOutput
+	// readIndex is the index of the next byte to read from the Output
+	readIndex int64
+	isClosed  bool
+}
+
+func NewOutputReadCloser(output *CommandOutput) *OutputReadCloser {
+	return &OutputReadCloser{output: output, readIndex: 0}
+}
+
+// Read reads from the Output and returns the number of bytes read and an error if any.
+//
+//	Wait for changes to the CommandOutput if no content is available to read.
+//	Returns EOF if the CommandOutput is closed and all the content has been read.
+func (outputReadCloser *OutputReadCloser) Read(buffer []byte) (n int, err error) {
+	if outputReadCloser.isClosed {
+		return 0, ErrReaderClosed
+	}
```

Comment:
In this design the Read and Close calls are not thread safe. 

So let say that consumer want to create a read loop that will block by r.Read call this a new output is available but  want to cancel the reeder loop 
```go 
ctx, cancel := ctx.WithCancel(ctx)
defer cancel()

r : NewOutputReadCloser()

go func() {
   switch {
    case <- ctx.Done:
        r.Close()
   } 
}
for  { 

  .., ... = r.Read(buff)
  ... 
}

```

So the question is how a consumer of outputReadCloser can cancel ongoing stream in the Read/Close function are not thread safe ?

Reply:
@smallinsky  Looks like RWMutex are best solutions in this case. 
<img width="718" alt="Screenshot 2024-09-05 at 6 32 07 PM" src="https://github.com/user-attachments/assets/294376ac-dd80-4227-a71a-7fe6830dd5a2">

And just to prove it -  I've created one more unit test ([Test_OutputReadCloser_all_calls_are_thread_safe](https://github.com/P-A-R-U-S/Go-Job-Worker-Service/blob/7d875c186cbd50bfc516d255adb26edd04ddf2ee/pkg/jobWorker/outputReadCloser_test.go#L156)):

1. Created Cancel context
2. Run go-routine with infinite loop to write output 
3. Run go-routine to receive Cancel status 
4. Run go-routine to trigger Cancel
5. Run infinite loop in main function to read output.
6. Expected: Read should stop reading data after calling Close function.

As you can see on sreenshot below after receiving `cancel` both reader stop processing data.
![Screenshot 2024-09-05 at 6 45 52 PM](https://github.com/user-attachments/assets/16697bab-497d-45bb-9bc2-1462308433a7)